base/array.jl

```diff
@@ -2730,7 +2730,7 @@ keepat!(a::Vector, m::AbstractVector{Bool}) = _keepat!(a, m)
 # set-like operators for vectors
 # These are moderately efficient, preserve order, and remove dupes.
 
-_unique_filter!(pred, update!, state) = function (x)
+_unique_filter!(pred::P, update!::U, state) where {P,U} = function (x)
```

Comment:
I remember running into a case once where an argument wouldn't specialize in a really complex inference case unless it was parameterized; is that what's going on here too?

Reply:
Yep. Since this is an internal function and it looks like the number of call combinations is limited, I think we can afford to force specialization. `map`, `filter`, etc. are obviously much more problematic.

Reply:
Perhaps it would be better to use the @specialize annotation. I'm not confident that it would work, but the approach of "adding type parameter to force specialization" is not very robust, as we may change the handling of (unmatched) type parameters in the future, as we did in https://github.com/JuliaLang/julia/pull/45062.

Reply:
Forgot that works now! Thanks!

Reply:
I am pretty sure `@specialize` just reverts to the default handling, and will never stop something like https://github.com/JuliaLang/julia/pull/45062 (since it is a compiler macro, not an inference macro)

Reply:
Sounds like something we should fix---we need a clear way of requesting an override of the heuristics. I'll leave it `@specialize` for now and we can check again when #47184 merges (too annoying to keep combining branches...).

Reply:
`::F` is the way to specify it. If we inline because we determine the optimizer computes the cost is free, that is fine. `@specialize` does not restrict that either, but also does less to ensure that.

Reply:
OK, I'll back out those `@nospecialize` changes

Reply:
Thanks for your comment @vtjnash .

> `::F` is the way to specify it.

Okay, so how does declaring an argument with a type variable change the inference behavior? Is it because of some implementation detail of `abstract_call_method`, e.g. the handling of `specTypes` is stricter than that of `sparam_vals` or something like that?

Reply:
`::F` does not change inference, nor does `@specialize`, but they both do impact caching computations

Reply:
Well, then how can `::F` prevent invalidations? If inference behaves the same, then we will compute the same backedge graph.